laplace/marglik_training.py

```diff
@@ -0,0 +1,259 @@
+
+    # get device, data set size N, number of layers H, number of parameters P
+    device = parameters_to_vector(model.parameters()).device
+    N = len(train_loader.dataset)
+    H = len(list(model.parameters()))
+    P = len(parameters_to_vector(model.parameters()))
+
+    # differentiable hyperparameters
+    hyperparameters = list()
+    # prior precision
+    log_prior_prec_init = np.log(temperature * prior_prec_init)
+    if prior_structure == 'scalar':
+        log_prior_prec = log_prior_prec_init * torch.ones(1, device=device)
+    elif prior_structure == 'layerwise':
+        log_prior_prec = log_prior_prec_init * torch.ones(H, device=device)
+    elif prior_structure == 'diagonal':
+        log_prior_prec = log_prior_prec_init * torch.ones(P, device=device)
+    else:
+        raise ValueError(f'Invalid prior structure {prior_structure}')
+    log_prior_prec.requires_grad = True
+    hyperparameters.append(log_prior_prec)
+
+    # set up loss (and observation noise hyperparam)
+    if likelihood == 'classification':
+        criterion = CrossEntropyLoss(reduction='mean')
+        sigma_noise = 1.
+    elif likelihood == 'regression':
+        criterion = MSELoss(reduction='mean')
+        log_sigma_noise_init = np.log(sigma_noise_init)
+        log_sigma_noise = log_sigma_noise_init * torch.ones(1, device=device)
+        log_sigma_noise.requires_grad = True
+        hyperparameters.append(log_sigma_noise)
+
+    # set up model optimizer
+    if optimizer_kwargs is None:
+        optimizer_kwargs = dict()
+    optimizer = optimizer_cls(model.parameters(), **optimizer_kwargs)
+
+    # set up learning rate scheduler
+    if scheduler_cls is not None:
+        if scheduler_kwargs is None:
+            scheduler_kwargs = dict()
+        scheduler = scheduler_cls(optimizer, **scheduler_kwargs)
+
+    # set up hyperparameter optimizer
+    hyper_optimizer = Adam(hyperparameters, lr=lr_hyp)
+
+    best_marglik = np.inf
+    best_model_dict = None
+    best_precision = None
+    losses = list()
+    margliks = list()
+
+    for epoch in range(1, n_epochs + 1):
+        epoch_loss = 0
+        epoch_perf = 0
+        
+        # standard NN training per batch
+        for X, y in train_loader:
+            X, y = X.to(device), y.to(device)
+            optimizer.zero_grad()
+            if likelihood == 'regression':
+                sigma_noise = torch.exp(log_sigma_noise).detach()
+                crit_factor = temperature / (2 * sigma_noise.square())
+            else:
+                crit_factor = temperature
+            prior_prec = torch.exp(log_prior_prec).detach()
+            theta = parameters_to_vector(model.parameters())
+            delta = expand_prior_precision(prior_prec, model)
+            f = model(X)
+            loss = criterion(f, y) + (0.5 * (delta * theta) @ theta) / N / crit_factor
+            loss.backward()
+            optimizer.step()
+            epoch_loss += loss.cpu().item() / len(train_loader)
```

Comment:
If I'm not missing something you now devide the loss by `N` in line 202, but you still have to multiply by `len(X)` here instead of divide by `len(train_loader)`.

Reply:
Oh yes, thanks. I mixed it up with `epoch_perf`...